compiler/rustc_lint/src/types.rs

```diff
@@ -884,9 +884,33 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
             if let Some(field) = transparent_newtype_field(self.cx.tcx, variant) {
                 self.check_field_type_for_ffi(cache, field, substs)
             } else {
-                // All fields are ZSTs; this means that the type should behave
-                // like (), which is FFI-unsafe
+                // All fields are ZSTs; this means that except if all
+                // fields are PhantomData the type should behave like (),
+                // which is FFI-unsafe,
+                let mut all_phantom = !variant.fields.is_empty();
+                for field in &variant.fields {
+                    let field_ty = field.ty(self.cx.tcx, substs);
+                    let field_ty = if field_ty.has_opaque_types() {
+                        field_ty
+                    } else {
+                        self.cx.tcx.normalize_erasing_regions(self.cx.param_env, field_ty)
+                    };
+                    match field_ty.kind() {
```

Comment:
It's better to follow what happens below where `check_field_type_for_ffi` is invoked and an `FfiPhantom` result is checked for, rather than what you're doing here.